### PR TITLE
Add "hotspot_2d" clustering algorithm

### PR DESCRIPTION
**Pull request by spenczar**
This PR adds a new algorithm for clustering points as an alternative to `dbscan`. I've nicknamed it "hotspot_2d" because it works by quantizing the X-Y points and then trying to find "hot" quantized X-Y pairs.

### The basic idea

The general idea is that we round the floating point X and Y values. Then, a "cluster" is just a series of these rounded values that appears often in the dataset.

To find values that appear often (say, at least N times), we first sort the dataset (by X and then Y), and then iterate through the sorted array. If we see the same X-Y pair N times consecutively, then we have found a cluster (internally, this is called a "run" in the code).

We have the clusters, but we don't know the indexes of the original, un-sorted X-Y values for each cluster, so then we do a final pass: we loop through the rounded-but-unsorted dataset, and check each value to see which "run" it belongs to, with a nested loop.

### Dealing with edge effects

At that point, we have generated cluster IDs for every datapoint, but we may have missed many clusters: some X-Y values will just unfortunately straddle our rounding resolution. For example, if we were rounding to the nearest 0.1, then we wouldn't recognize `[(1.04, 2.04), (1.04, 2.04), (1.06, 2.06)]` as a cluster of size 3, because it would round to `[(1.0, 2.0), (1.0, 2.0), (1.1, 2.1)]`

To deal with this, we jitter the datapoints a bit, and re-run the clustering algorithm. We do this 3 times, each time offsetting by the half the rounding scale (call it `eps`). First we offset by X+eps/2, then Y+eps/2, then both. So, in that previous example, we would jitter the points to look like

 - x+eps/2:   `[(1.09, 2.04), (1.09, 2.04), (1.11, 2.06)]` -> `[(1.1, 2.0), (1.1, 2.0), (1.1, 2.1)]`: ❌
 - y+eps/2:   `[(1.04, 2.09), (1.04, 2.09), (1.06, 2.11)]` -> `[(1.0, 2.1), (1.0, 2.1), (1.1, 2.1)]`: ❌
 - x+eps/2, y+eps/2:   `[(1.09, 2.09), (1.09, 2.09), (1.11, 2.11)]` -> `[(1.1, 2.1), (1.1, 2.1), (1.1, 2.1)]`: ✔️

Now, we have 4 indexes of cluster IDs associated with each point. We need to coalesce them together. We do this by taking two passes: first, walk through all 4 indexes, and build up a dictionary of "aliases", for cases where the same point appears in multiple indexes. Then, walk through the points and label them using those aliases.

This has some buggy, weird behavior if there is ever a full cluster in one of the indexes which is split into _two_ clusters in a different index; my hope is that clusters should rarely be that big.

### Performance

I mostly measured this on my laptop, using the benchmark dataset committed in `bench/data`. I used `%timeit` mostly, with a bit of `cProfile` profiling, although `cProfile` isn't able to dig in through numba's compiled bytecode, so this was not so useful. I generally get times around 35ms for cluster-finding. I think that, with some effort, we could get that down to 30ms or even 25ms, if that's worth it, but I am not sure it will be.

Increasing `eps` makes the algorithm slower, as more clusters get found and need to be coalesced. Increasing `min_samples` makes the algorithm faster, as fewer runs get identified. Increasing the number of points makes things slower, approximately linearly.

### Accuracy

I have not done any careful tests for accuracy or completeness of the clusters this generates. I just kind of eyeballed that it gives similar answers to `dbscan`. It tends to find a few more clusters than `dbscan`, perhaps 10-20% more, but I am not too sure about that.

### CODE REVIEW

File: bench/benchmark_cluster.py [hunk 1]
```diff
@@ -1,8 +1,11 @@
 from thor import clusterVelocity
```
Comment (spenczar) [inline bench/benchmark_cluster.py:1]: Feel free to review this file closely, but I don't think it's very important. I just rearranged things to make it generate a lot of profiles for different parameter values (like different numbers of datapoints, `eps`, `min_samples`). It's semi-useful; a bit verbose and the result is hard to read.

File: thor/main.py [hunk 2]
```diff
@@ -49,12 +51,12 @@
 ]
 
 def rangeAndShift_worker(observations, ephemeris, cell_area=10):
-
```
Comment (spenczar) [inline thor/main.py:52]: Sorry about all the whitespace changes! My emacs setup cleans up whitespace on save. Maybe I should write a script to go through the whole repo so we don't constantly look at this crap.

Comment (moeyensj) [reply thor/main.py:52]: No worries, I updated my vscode settings to do the same. I'll take care of the trailing whitespaces in a single commit after I merge this.

**Comment by moeyensj** [comment]
I'm pretty excited to try this algorithm and see the effect on completeness and the speed gains. @spenczar I'm happy to merge if you feel like this PR is ready to go. 

cc' @mjuric

**Comment by spenczar** [comment]
@moeyensj Sounds good, feel free to merge. I'm really interested to hear how it turns out. I'm cautiously optimistic!

**Comment by spenczar** [comment]
By the way - coveralls complains because it can't tell which lines have been touched when numba compiles a function. We could run tests with an environment variable set: `NUMBA_DISABLE_JIT=1` will turn it all off. We'd gain test coverage numbers, but lose a bit of accuracy.

We could run tests twice in CI, once with JIT and once without, to generate correct coverage maps... I'm not sure.

**Comment by mjuric** [comment]
@spenczar This looks awesome!!!

I didn't have the time to look at the code, but if there's a way to selectively enable/disable the new algorithm that would be ideal. So we can know which differences come because of the new clustering algo, and which ones because of Ray snafus.

**Comment by spenczar** [comment]
@mjuric Yep, that should be possible. It was added in https://github.com/moeyensj/thor/pull/52 - there was just only one valid value of `alg`, which was `"dbscan"`. Now `"hotspot_2d"` is the default, but you can set `alg="dbscan"` instead too.

There's no entrypoint from the outside, though, like a config file or a command-line flag. You'd have to edit the source code, albeit just in one spot. Hopefully that should be easy to improve to match any way we'd like to compare things.

**Pull request merged**